### DISCUSSION TITLE
The @ Error Control Operator is no longer suppressing errors in Blade templates

### DISCUSSION THREAD

**Discussion by warmwhisky** [laravel/framework#44116] [Q&A]
- Laravel Version: 9.29.0
- PHP Version: 8.1
- Database Driver & Version: Mysql 8

### Description:
The @ Error Control Operator is no longer suppressing errors in Blade templates

### Steps To Reproduce:
Use the @ sign before a variable. `<input type="hidden" name="user_id" value="{{ @$user->id }}">`
I could change it to `{{ isset($user) ? $user->id : '' }}` but wondered if something had changed or been deprecated?

**Comment by rodrigopedra** [comment]
Can you share the error? Just tested on a fresh install just changing the `welcome.blade.php` template to:

~~~blade
<input type="hidden" name="user_id" value="{{ @$user->id }}">
~~~

and no error was thrown and the output is as expected (an `<input>` tag with empty `value` attribute).

As I didn't pass any variables to the view, it should error by no `$user` variable been defined, right?

Also, as a tip, since PHP 8 you can do this instead:

~~~blade
<input type="hidden" name="user_id" value="{{ $user?->id }}">
~~~

which, in my opinion, is clearer.

**Reply by warmwhisky** [reply]
Thanks for getting back to me.

I will share the error tomorrow when I am back at work.

I didn't know about `{{ $user?->id }}` in PHP 8! It is far more clear. I will be using this instead of the `@` sign from now on. Thanks for the tip

**Reply by warmwhisky** [reply]
@rodrigopedra Here is a link to the error https://flareapp.io/share/17D2lkzm#F64

**Reply by rodrigopedra** [reply]
I tried again, setting `$user` to `null` and still cannot reproduce it with my default settings.

Do you have xdebug extension installed? If so what is the value of `xdebug.scream` value?

You can search it with this command:

~~~shell
$ php -i | grep xdebug.scream
~~~

Searching for ways to disable the error suppression operator, setting `xdebug.scream` to `On` was the only way I could reproduce the error.

If you have set to `On`, consider setting it to `Off` on production. Or at least try to cacth any `ErrorException` and report internally without braking the app.

Actually I think it is better to not have xdebug enabled in production to enhance performance (but I can be wrong on this last statement).

**Reply by rodrigopedra** [reply]
StackOverflow answer about having xdebug enable on production: https://stackoverflow.com/a/3522356

**Reply by warmwhisky** [reply]
I tried

```
$ php -i | grep xdebug.scream
```
And got this error

```
PHP Warning:  PHP Startup: Unable to load dynamic library 'mysqli' (tried: /usr/lib/php/20210902/mysqli (/usr/lib/php/20210902/mysqli: cannot open shared object file: No such file or directory), /usr/lib/php/20210902/mysqli.so (/usr/lib/php/20210902/mysqli.so: undefined symbol: mysqlnd_global_stats)) in Unknown on line 0

```
This is a fresh install of Ubuntu 22.04. Not sure if that is something to do with it. I have not installed x debug and from brief search I don't believe it comes preinstalled on Ubuntu 22.04. I am running Laravel with Docker. I tried the following
```
sail php -i | grep xdebug.scream
```
and got 

```
xdebug.scream => Off => Off
```

After I changed `@$user->id` to `$user?->id` it got past that line but it failed on the next one 
```
<input type="hidden" name="dhl_zone" value="{{ $shippingAddress?->dhl_zone }}">
```

**Reply by rodrigopedra** [reply]
as

~~~
sail php -i | grep xdebug.scream
~~~

returns something, even if its  is `Off` it means you have the xdebug installed. Sail uses docker under the hood, so some PHP extensions are installed on the docker image/volume.

Maybe there is some other extension or `php.ini` config sail uses that affects the error suppression operator. 

As I don't use sail or docker, I can't tell which. 

But on a plain install no errors are triggered.

About the next error, on the shipping address, is the error message the same?

**Reply by warmwhisky** [reply]
The shipping error is not the same. it says `Undefined variable $shippingAddress` [https://flareapp.io/share/NPG6zgNm#F64](url)

I also added the following to php.ini in ./docker/8.1 and restarted the container to no avail
```
xdebug.remote_autostart=0  
xdebug.remote_enable=0
xdebug.profiler_enable=0
```

**Reply by rodrigopedra** [reply]
well, are you passing any variable named `$shippingAddress` to the view?

In your case, from the snippet on flare, as you have a `if` clause to test for `$shippingAddress` presence you can change this part of your view from:

~~~blade
<input type="hidden" name="dhl_zone" value="{{ $shippingAddress?->dhl_zone }}">
@if(!isset($shippingAddress))
    <input type="hidden" name="new_shipping_address" value="1">
@endif
~~~

To: 

~~~blade
@isset($shippingAddress)
    <input type="hidden" name="dhl_zone" value="{{ $shippingAddress->dhl_zone }}">
@else
    <input type="hidden" name="new_shipping_address" value="1">
@endisset
~~~

Note I changed from using `@if` to `@isset` which is a built-in blade directive.

Reference: https://laravel.com/docs/9.x/blade#if-statements

**Reply by warmwhisky** [reply]
There is no variable `$shippingAddress` being passed in this case as its a logged out user.
However, this has worked fine for 2 years and also works fine with the same repository in Windows 10 environment.

Using your `@isset()` works. Thanks for another tip there.

Then it failed on the next usage or `@`

```blade
@foreach($countries as $country)
    @if(@$shippingAddress->delivery_country_code == $country->alpha_2_code)
        <option value="{{$country->alpha_2_code}}" selected>
            {{$country->en_short_name}}
        </option>
    @else
        <option value="{{$country->alpha_2_code}}">
            {{$country->en_short_name}}
        </option>
    @endif
@endforeach
```
I have tried to uninstall xdebug from Ubuntu, but apparently its not installed.
I don't see why all of a sudden `@` no longer works for variables that do not exist?

**Reply by rodrigopedra** [reply]
Well, I tried running sail locally and I can't still reproduce the error. Might be something specific to your install. You can try reinstalling sail on this project. Or try in a different machine if you have one available.

If it still fails, I guess you will have to assess error per error... and I unfortunately can't help with all of them.

On this last snippet I would refactor it to use the new-ish `@selected()` blade directive:

~~~blade
@foreach($countries as $country)
    <option value="{{$country->alpha_2_code}}" @selected($shippingAddress?->delivery_country_code == $country->alpha_2_code)>
        {{$country->en_short_name}}
    </option>
@endforeach
~~~

https://laravel.com/docs/9.x/blade#additional-attributes

Sorry for not founding the root cause. It is working fine here. Good luck =)

**Reply by warmwhisky** [reply]
Thanks for all of your input and for all of the new blade directives. I will try reinstall Sail and see what happens.

I am back on my Windows drive now and tested exactly the same environment and there is no error :/

**Reply by rodrigopedra** [reply]
You're welcome! My pleasure to help. Have a nice day =)

**Reply by warmwhisky** [reply]
On Monday, I first tried Ubuntu using a repo i'd been using in Windows 10 and Docker. Initially I had a lot of issues getting the correct Meilisearch version with docker and may have caused an issue with updating some of the packages. Its a maybe, but with a fresh repo install and knowing how to get the correct version of Meilisearch installed I no longer have the issues with `@` and I have a few new blade directives to work with. Thanks!

**Reply by warmwhisky** [reply]
Well that did not last long. After a reboot I am back to `@` not working again.

Something very strange is going on here.
